File: java/anagram/src/main/java/Anagram.java
import java.util.List;
import java.util.stream.Collectors;

public class Anagram {
    private final String given;
    private final String givenSorted;

    public Anagram(String given) {
        this.given = given.toLowerCase();
        this.givenSorted = sort(this.given);
    }

    private static String sort(String s) {
        return s.chars()
                .sorted()
                .mapToObj(i-> String.valueOf((char)i))
                .collect(Collectors.joining());
    }

    public List<String> match(List<String> candidates) {
        return candidates.stream()
                .filter(candidate -> !given.equals(candidate.toLowerCase()))
                .filter(candidate -> givenSorted.equals(sort(candidate.toLowerCase())))
                .collect(Collectors.toList());
    }
}

File: haskell/anagram/src/Anagram.hs
module Anagram (anagramsFor) where

import qualified Data.List as L
import qualified Data.Char as C

anagramsFor :: String -> [String] -> [String]
anagramsFor xs = filter isAnagram
  where
    lower = toLowers xs
    sorted = L.sort lower
    toLowers = map C.toLower
    isAnagram zs
      | lower /= toLowers zs &&
        sorted == (L.sort . toLowers) zs = True
      | otherwise = False